### PR TITLE
Alert context appsec

### LINKED ISSUE #3286: [appsec / parsers] Add expr helper or event function that can add custom Alert context at runtime
Currently if you want to add Alert context you can only do this from events that generate from a overflown scenario.

However, some alerts get sent without an overflown scenario (AppSec), the idea is to add either an expr or a event level function that adds alert context metadata.

The function will need to check if the key already exists and push to the slice.

### PR DESCRIPTION
**Pull request by buixor**
Add alert context support to appsec rules. Closes #3286

When we release 1.6.4, we should add the following alert context to the appsec collection(s) to be iso with what exists:

```yaml
context:
  id:
   - match.id
  name:
   - match.name
  method:
   - match.method
  uri:
   - match.uri
  matched_zones:
   - match.matched_zones
  msg:
   - match.msg
```

However, we're now exposing both `match` (`types.MatchedRule`) and `req` (`http.Request`) to the user's alert context rules. This allows extracting more relevant context:

```yaml
context:
#extract the content of specific header
  smth:
     - req.Header.Get("foobar")
 #extract all cookies
  cookies:
     - req.Cookies()
  #parse the cookie and extract a specific one
  foobar_cookie:
     - find(req.Cookies(), {#.Name == "yop"}).String()

```

Todo:
 - [x] Coverage

### CODE REVIEW

**Comment by buixor** [comment]
todo: use expr lib

**Comment by LaurenceJJones** [comment]
Testing PR:

Added contents to `appsec_base.yaml` provided via the hub

```
root@bookworm:/opt/crowdsec# cat /etc/crowdsec/contexts/appsec_base.yaml
context:
  test_header:
   - req.Header.Get("x-test")
  id:
   - match.id
  name:
   - match.name
  method:
   - match.method
  uri:
   - match.uri
  matched_zones:
   - match.matched_zones
  msg:
   - match.msg
  rules:
    - evt.Meta.rule_name
  target_uri:
    - evt.Meta.target_uri
```
produces the following context with the request `curl -H "X-Test: mytest" http://192.168.121.18/.env`

```
 - Context  :
╭───────────────┬─────────────────────────────────╮
│      Key      │              Value              │
├───────────────┼─────────────────────────────────┤
│ id            │ 3256131465                      │
│ matched_zones │ REQUEST_FILENAME                │
│ method        │ GET                             │
│ msg           │ crowdsecurity/vpatch-env-access │
│ name          │ crowdsecurity/vpatch-env-access │
│ test_header   │ mytest                          │
│ uri           │ /.env                           │
╰───────────────┴─────────────────────────────────╯
```

Some things to note is the `msg` and `name` seem the be the exact same so we might want to only provide one or the other depending if custom rules will no produce the same duplicate effect.

To test directly the reason for the PR I added the following to the context file:

```
context:
  host:
   - req.Host
```

Sending these 2 curl commands:
```
curl -H "X-Test: mytest" http://192.168.121.18/.env -vv
curl -H "X-Test: mytest" -H "Host: test.example.com" http://192.168.121.18/.env -vv
```

Produces the following contexts:
```
root@bookworm:/opt/crowdsec# cscli alerts inspect 6

################################################################################################

 - ID           : 6
 - Date         : 2024-10-20T08:41:28Z
 - Machine      : c1a6b6ec59fa4f10977727267fd20e4d
 - Simulation   : false
 - Remediation  : false
 - Reason       : crowdsecurity/vpatch-env-access
 - Events Count : 0
 - Scope:Value  : Ip:192.168.121.1
 - Country      :
 - AS           :
 - Begin        : 2024-10-20 08:41:28 +0000 UTC
 - End          : 2024-10-20 08:41:28 +0000 UTC
 - UUID         : bfd99f7b-a806-452e-be42-597d7eb3b6c9


 - Context  :
╭───────────────┬─────────────────────────────────╮
│      Key      │              Value              │
├───────────────┼─────────────────────────────────┤
│ host          │ 192.168.121.18                  │
│ id            │ 3256131465                      │
│ matched_zones │ REQUEST_FILENAME                │
│ method        │ GET                             │
│ msg           │ crowdsecurity/vpatch-env-access │
│ name          │ crowdsecurity/vpatch-env-access │
│ test_header   │ mytest                          │
│ uri           │ /.env                           │
╰───────────────┴─────────────────────────────────╯
root@bookworm:/opt/crowdsec# cscli alerts inspect 7

################################################################################################

 - ID           : 7
 - Date         : 2024-10-20T08:41:51Z
 - Machine      : c1a6b6ec59fa4f10977727267fd20e4d
 - Simulation   : false
 - Remediation  : false
 - Reason       : crowdsecurity/vpatch-env-access
 - Events Count : 0
 - Scope:Value  : Ip:192.168.121.1
 - Country      :
 - AS           :
 - Begin        : 2024-10-20 08:41:51 +0000 UTC
 - End          : 2024-10-20 08:41:51 +0000 UTC
 - UUID         : b8ab4385-5c5f-41c8-be6e-462b6eac04ca


 - Context  :
╭───────────────┬─────────────────────────────────╮
│      Key      │              Value              │
├───────────────┼─────────────────────────────────┤
│ host          │ test.example.com                │
│ id            │ 3256131465                      │
│ matched_zones │ REQUEST_FILENAME                │
│ method        │ GET                             │
│ msg           │ crowdsecurity/vpatch-env-access │
│ name          │ crowdsecurity/vpatch-env-access │
│ test_header   │ mytest                          │
│ uri           │ /.env                           │
╰───────────────┴─────────────────────────────────╯
```

**Comment by buixor** [comment]
:+1: for keeping only `name` or `msg`. Is your statement valid too for modsec native rules ?

Todo:
 - [ ] Create the AC PR for appsec

**Review by LaurenceJJones — APPROVED** [review]

File: pkg/types/event.go [hunk 1]
```diff
@@ -47,6 +47,13 @@ type Event struct {
 	Meta map[string]string `yaml:"Meta,omitempty" json:"Meta,omitempty"`
 }
 
+func EmptyEvent() *Event {
+	return &Event{Type: LOG,
+		Parsed:   make(map[string]string),
+		Enriched: make(map[string]string),
+		Meta:     make(map[string]string)}
```
Comment (blotus) [inline pkg/types/event.go:54]: Missing `Unmarshaled: make(map[string]interface{})`

File: pkg/types/event.go [hunk 2]
```diff
@@ -47,6 +47,13 @@ type Event struct {
 	Meta map[string]string `yaml:"Meta,omitempty" json:"Meta,omitempty"`
 }
 
+func EmptyEvent() *Event {
```
Comment (blotus) [inline pkg/types/event.go:50]: Maybe rename to `NewEvent` ?

**Review by blotus — APPROVED** [review]

**Pull request merged**